### PR TITLE
split steps of initialization of `terminal` into chains of base class

### PR DESCRIPTION
**Pull request by ImpleLee**
This is more like a proof of concept to move steps of construction and destruction into chains of base class, so each completed initialization will trigger the corresponding destruction when there is any exception etc, thus eliminating the need of `defer`.

When there are multiple steps of initialization, `defer` objects should be made either a class member to remember the ownership (what I did in the first commit), or a stack object used with careful `release` (the original way), but the latter I think is like an emulation of the former... Making these steps chains of base class is just another way to make them sub-objects, and in this way they can also share class members...

### CODE REVIEW

File: view/cli/src/terminal.cpp [hunk 1]
```diff
@@ -191,13 +195,10 @@ void terminal::append(const char* s, size_t length) {
...
 
 void terminal::flush() {
-	if(write(term, buffer.data(), buffer.size()) != buffer.size()) {
-		std::stringstream error;
-		error << "cannot write to terminal: " << strerror(errno);
-		throw std::runtime_error(error.str());
-	} else {
-		buffer = std::vector<char>();
-	}
+	write_term(term, std::string{buffer.begin(), buffer.end()}, []{
```
Comment (aegistudio) [inline view/cli/src/terminal.cpp:198]: Unintentionally constructed string, with memory allocation and string copying.

File: view/cli/src/terminal.cpp [hunk 2]
```diff
@@ -20,30 +20,38 @@
 #include <sstream>
 #include <cstring>
 #include <stdexcept>
+#include <functional>
+
+static void write_term(int term, const std::string &content, const std::function<std::string()> &error_message) {
```
Comment (aegistudio) [inline view/cli/src/terminal.cpp:25]: We don't require it to be so elastic as receiving a lambda function for constructing the error message. For us, it is enough just to annotate the operation context, represented by certain string.

File: view/cli/include/view/cli/terminal.hpp [hunk 3]
```diff
@@ -88,6 +88,22 @@ struct pos {
 	constexpr pos(int x, int y): x(x), y(y) {}
 };
 
+namespace details {
+struct initialized_terminal {
```
Comment (aegistudio) [inline view/cli/include/view/cli/terminal.hpp:92]: Enforced camel case.

**Comment by ImpleLee** [comment]
see #3 for the true branch to be merged... or to be short, I failed to force push...

**Pull request closed without merging**